heroic-elasticsearch-utils/src/test/java/com/spotify/heroic/elasticsearch/ScrollTransformTest.java

```diff
@@ -0,0 +1,167 @@
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+import static org.mockito.Mockito.any;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class ScrollTransformTest {
+    @Mock
+    private AsyncFramework async;
+
+    @Mock
+    private AsyncFuture<LimitedSet> resolved;
+
+    @Mock
+    Supplier<AsyncFuture<SearchResponse>> scroller;
+
+    @Mock
+    AsyncFuture<SearchResponse> response;
+
+    @Mock
+    SearchResponse searchResponse;
+
+    @Mock
+    SearchHits searchHits;
+
+    private final SearchHit[] searchHits1 = {
+        mock(SearchHit.class),
+        mock(SearchHit.class),
+        mock(SearchHit.class),
+    };
+
+    private final SearchHit[] searchHits2 = {
+        mock(SearchHit.class),
+        mock(SearchHit.class),
+        mock(SearchHit.class),
+    };
+
+    private final SearchHit[] emptySearchHits = {};
+
+    @Before
+    public void setup() {
+        doReturn(resolved).when(async).resolved(any(LimitedSet.class));
+        doReturn(searchHits).when(searchResponse).getHits();
+
+        doReturn(response).when(scroller).get();
+        doAnswer(new Answer<AsyncFuture<LimitedSet<Integer>>>() {
+            public AsyncFuture<LimitedSet<Integer>> answer(
+                InvocationOnMock invocation
+            ) throws Exception {
+                LazyTransform<SearchResponse, LimitedSet<Integer>> transform =
+                    (LazyTransform<SearchResponse, LimitedSet<Integer>>)
+                        invocation.getArguments()[0];
+                return transform.transform(searchResponse);
+            }
+        }).when(response).lazyTransform(any(LazyTransform.class));
+    }
+
+    public void setSearchHitPages(SearchHit[]... pages) {
+        OngoingStubbing<SearchHit[]> stub = when(searchHits.getHits());
+        for (final SearchHit[] page : pages) {
+            stub = stub.thenReturn(page);
+        }
+    }
+
+    public ScrollTransform<Integer> createScrollTransform(
+        Integer limit,
+        Supplier<AsyncFuture<SearchResponse>> scroller
+    ){
+        final OptionalLimit optionalLimit = OptionalLimit.of(limit);
+        return new ScrollTransform<>(async, optionalLimit, scroller, Object::hashCode);
+    }
+
+    public LimitedSet<Integer> createLimitSet(Integer limit, SearchHit[]... pages){
+        final LimitedSet<Integer> limitedSet =
+            new LimitedSet<>(new HashSet<>(), limit != null);
+        Stream<SearchHit> stream = Arrays.stream(pages)
+              .map(Arrays::stream)
+              .reduce(Stream.empty(), Stream::concat);
+        if (limit!=null){
+            stream = stream.limit(limit);
+        }
+        stream.map(Object::hashCode)
+              .forEach(limitedSet.getSet()::add);
```

Comment:
Containers are generally designed to be immutable. I don't consider it a safe assumption that the set returned by this getter is necessarily the same (it happens to be now) or that it is mutable. It also looks like it should be possible to build the `HashSet` first, then constructing a `LimitedSet` out of that.